### DISCUSSION TITLE
Firebase Authentication

### DISCUSSION THREAD

**Discussion by transmissions11** [vercel/next.js#13016] [Help] [answered]
Hey all, 

Really want to get into building apps with Next.js, but was really confused when I hit the example with Firebase Authentication, as it seems far overly complex. Is there really no simpler way to use SSR with Firebase Auth then having to create signIn endpoints, etc?

Or are there any libraries that simplify this process?

**Comment by brianlovin** [comment]
I'm a bit confused by your question – if you want authentication, you'd probably need ways to sign in, sign out, and register, right? 

In case I misunderstood that bit, another option is to look at services like Auth0 (https://auth0.com). They have some tutorials (https://auth0.com/blog/ultimate-guide-nextjs-authentication-auth0/) for how to integrate with Next.js that might make your life a bit easier.

**Comment by rafinskipg** [comment, marked as answer]
Firebase authentication works on the client-side too, you do not need API routes, just do as the firebase documentation says. https://firebase.google.com/docs/auth/web/firebaseui

**Reply by rafinskipg** [reply]
API authentication offers a more fine-grained control, but it requires more time to develop

**Reply by rafinskipg** [reply]
If you want the server to render authenticated routes, then you will need to rely in a cookie (I think it is the only way to know if a user is authenticated when visiting the page) or another sort of "built-in" identification mechanism on the browser.

**Reply by tettoffensive** [reply]
What do you mean by more fine-grained control? does it offer more security? better user experience? i too am trying to understand.

**Reply by francisrod01** [reply]
@rafinskipg is this way safe enough?
I'm trying to combine the two examples using redux but I can't see reading data from a cookie on SSR be a safe way.
- https://github.com/vercel/next.js/blob/canary/examples/with-firebase-authentication
- https://github.com/vercel/next.js/blob/canary/examples/with-redux-thunk
